src/rust/ensogl/src/application/shortcut.rs

```diff
@@ -25,46 +28,66 @@ use crate::frp;
 /// shortcut manager which will own and manage the handles.
 #[derive(Clone,CloneRef,Debug)]
 pub struct Registry {
+    model   : RegistryModel,
+    network : frp::Network,
+}
+
+/// Internal representation of `Registry`.
+#[derive(Clone,CloneRef,Debug)]
+pub struct RegistryModel {
     logger            : Logger,
     keyboard          : Keyboard,
     keyboard_bindings : Rc<KeyboardFrpBindings>,
-    network           : frp::Network,
     command_registry  : command::Registry,
-    rule_map          : Rc<RefCell<HashMap<keyboard::KeyMask,Vec<Instance>>>>
+    action_map        : Rc<RefCell<ActionMap>>
 }
 
-impl Registry {
+impl Deref for Registry {
+    type Target = RegistryModel;
+    fn deref(&self) -> &Self::Target {
+        &self.model
+    }
+}
+
+
+impl RegistryModel {
     /// Constructor.
     pub fn new(logger:&Logger, command_registry:&command::Registry) -> Self {
         let logger            = logger.sub("ShortcutRegistry");
         let keyboard          = Keyboard::default();
         let keyboard_bindings = Rc::new(KeyboardFrpBindings::new(&logger,&keyboard));
         let command_registry  = command_registry.clone_ref();
-        let rule_map          = default();
-        let network           = default();
-        Self {logger,keyboard,keyboard_bindings,network,command_registry,rule_map} . init()
+        let action_map        = default();
+        Self {logger,keyboard,keyboard_bindings,command_registry,action_map}
+    }
+}
+
+impl Registry {
+    /// Constructor.
+    pub fn new(logger:&Logger, command_registry:&command::Registry) -> Self {
+        let model = RegistryModel::new(logger,command_registry);
+        frp::new_network! { network
+            eval model.keyboard.key_mask          ((m) model.process_action(ActionType::Press,m));
+            eval model.keyboard.previous_key_mask ((m) model.process_action(ActionType::Release,m));
```

Comment:
what is meaning of m?
And this code looks a little hacky, that the events of previous_key_mask do the ActionType::Release. In fact the previous_key_mask should be updated on every event, not only on releases.

Reply:
Oh, you are right!

Reply:
I will leave fixme for next PR here.

Reply:
Oh no, wait, it works fine :D